deploy: fix the owner reference kind to be rc

mfojtik opened the pull request:
This PR will:

* Fix the "kind" field in the ownerRef from deployerPod to point to RC instead of DC
* Automatically set ownerRef from lifecycle hook pods created by deployer pods to RC (so we clean them up when the RC is removed)
* Fix infra policy to use deployment SA instead of deploymentConfig SA in deployment controller
* Fix permissions on pods for deploymentConfig SA (don't need to patch, update, delete or create)
* Use the SA client in deploymentConfig controller instead of privileged client

mfojtik commented:
[test]

mfojtik commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -435,7 +436,9 @@ func (c *DeploymentConfigController) cleanupOldDeployments(existingDeployments [
 		}
 
-		err := c.rn.ReplicationControllers(deployment.Namespace).Delete(deployment.Name, nil)
+		err := c.rn.ReplicationControllers(deployment.Namespace).Delete(deployment.Name, &metav1.DeleteOptions{
```
@kargakis check.

deads2k commented on pkg/deploy/controller/deployment/controller.go:
```diff
@@ -329,10 +329,8 @@ func (c *DeploymentController) makeDeployerPod(deployment *kapi.ReplicationContr
-				// clientsets.
 				APIVersion: "v1",
-				Kind:       deployapi.Kind("DeploymentConfig").Kind,
+				Kind:       kapi.Kind("ReplicationController").Kind,
```
@ncdc any chance this is the source of your problem?  The names wouldn't match, right?

deads2k replied:
Actually, @mfojtik this pod is more related to the DC than the RC and should be cleaned up when the DC is deleted, right?  That would mean the kind was correct and the name and UID were wrong.

mfojtik replied:
it will be cleaned up because when you delete DC that should trigger deletion of RC and that should remove the deployer pod. Similarely when you set revisionHistory, this should automatically remove failed deployer pods (success deployer pods are removed automatically) as we delete old RC. correct @kargakis ?

kargakis replied:
@mfojtik correct, this pod is more related to the RC than the DC.

kargakis replied:
We also need to set finalizers in the RCs for enabling a cascading deletion from a DC to reach pods but that should be part of https://trello.com/c/B10xLQ6L/927-set-owner-refs-in-rcs-owned-by-dcs.

kargakis commented:
Needs a test that verifies GC of deployers works.

kargakis commented:
> Needs a test that verifies GC of deployers works.

@mfojtik I think you can re-use the existing test for cleaning up older RCs.

ncdc commented:
What I'd really like to see is a test that can reliably reproduce the transient deletion we saw in Jenkins that is then fixed by this PR (in whatever form it ends up in).

mfojtik commented:
@ncdc @kargakis @deads2k does kube GC even check the "kind" ? it seems like the UUID is correct (it points to the RC). Also i'm not able to reproduce the failure Andy saw... I tried:

1) Make pre-hook with sleep 120 (in theory the deployer pod should be erased prematurely by GC?)
2) Make pre-hook with abort policy and exit 1, which will cause deployment controller to keep the pod around (not removed by deployment controller)... it works as expected...
3) Same as 2) but set revisionHistory limit to 2, I see all the failed deployer pods and hook pods sitting around and they are never garbage collected (with this broken, I would expect they will be vanished as they point to an non-existent DC (the UUID is RC).

... unless the GC only checks the UUID and not the Kind. In that case the cause of the problem is somewhere else and this is just a cosmetic issue.

mfojtik commented:
OK, so it seems like the GC honors the UID from the owner reference, which means this is not fixing https://github.com/openshift/origin/issues/13995 but we should fix this anyway...

i'll add owner reference to hook pods in this PR.

kargakis commented:
Make sure that when we cleanup RCs based on [revisionHistoryLimit](https://github.com/openshift/origin/blob/13b90e25741d8c2d945ac313ca98f32eff04756a/pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go#L4380 we delete with propagationPolicy set to Background.

ncdc commented:
The failure that I originally found was from a CI run for #13886. It didn't look like that PR would cause the problem. Either way, I still haven't reasoned through how this could have happened. That's why I want a test case that can "slow down time" and try to be prescriptive about what happens and see if we can reproduce.

mfojtik commented:
@ncdc sure, I will work on that test, I think this PR is not related to that issue at this point.

@deads2k @liggitt @enj will you guys have any objections granting the "deployer" SA ability to delete pods in the project? When I add ownerRef in a hook pod created by the deployer I'm hitting this error:

```
 cannot set an ownerRef on a resource you can't delete: User "system:serviceaccount:test:deployer" cannot delete pods in project
```

There are two options we discussed with @kargakis:

1) Grant deployer delete
2) Update the failed hook pods (which are subject of GC when the RC is removed) with ownerRef from the controller (basically when a deployer controller see a failed hook pod, it automatically patches it with the ownerRef).

I like 1) because it is cleaner, but less secure, but will take an advise from you :)

mfojtik commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -998,7 +998,7 @@ func (c *MasterConfig) DeploymentConfigInstantiateClients() (*osclient.Client, k
 // DeploymentControllerClients returns the deployment controller client objects
 func (c *MasterConfig) DeploymentControllerClients() (*osclient.Client, kclientsetinternal.Interface, kclientsetexternal.Interface) {
-	_, osClient, internalKubeClientset, externalKubeClientset, err := c.GetServiceAccountClients(bootstrappolicy.InfraDeploymentConfigControllerServiceAccountName)
+	_, osClient, internalKubeClientset, externalKubeClientset, err := c.GetServiceAccountClients(bootstrappolicy.InfraDeploymentControllerServiceAccountName)
```
@deads2k fyi

mfojtik commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -208,7 +208,7 @@ func init() {
 				// DeploymentController.podClient
 				{
-					Verbs:     sets.NewString("get", "list", "create", "watch", "delete", "update"),
+					Verbs:     sets.NewString("get", "list", "watch"),
```
@deads2k down here we have:

```
 				{
 					Verbs:     sets.NewString("create", "update", "patch"),
 					Resources: sets.NewString("events"),
 				},
```

and this is copied in other policies... do we really need PATCH and UPDATE for events?

kargakis replied:
Maybe patching/updating events has to do with event compression?

enj commented:
> will you guys have any objections granting the "deployer" SA ability to delete pods in the project

@mfojtik I am missing some context on this PR but shouldn't you use a custom controller SA for this?  @deads2k's #14033 is working towards that.  Then you could grant that SA whatever permissions it needed.  Maybe there are other issues I am missing in regards to this being inside a project...

I cannot comment on if you actually have other viable solutions to the problem as a whole.

kargakis commented:
> @mfojtik I am missing some context on this PR but shouldn't you use a custom controller SA for this? @deads2k's #14033 is working towards that. Then you could grant that SA whatever permissions it needed. Maybe there are other issues I am missing in regards to this being inside a project...
> 
> I cannot comment on if you actually have other viable solutions to the problem as a whole.

We are already using a custom sa for the deployer, the question is "do we want to give pod deletion power to the deployer?"

kargakis commented on pkg/deploy/controller/deployment/deployment_controller.go:
```diff
@@ -414,9 +428,64 @@ func (c *DeploymentController) makeDeployerContainer(strategy *deployapi.Deploym
+	}
+
+	encoder := kapi.Codecs.LegacyCodec(kapi.Registry.EnabledVersions()...)
+	errors := []error{}
```
var errors []error

kargakis commented on pkg/deploy/controller/deployment/deployment_controller.go:
```diff
@@ -414,9 +428,64 @@ func (c *DeploymentController) makeDeployerContainer(strategy *deployapi.Deploym
+		}
+		newPod, ok := objCopy.(*kapi.Pod)
+		if !ok {
+			errors = append(errors, fmt.Errorf("object %#+v is not a pod", objCopy))
```
It's unlike that this error will ever be hit but if you want it, that's fine.

mfojtik replied:
@kargakis for the sake of somebody will copy&paste this code somewhere else :-)

enj commented:
> We are already using a custom sa for the deployer, the question is "do we want to give pod deletion power to the deployer?"

@kargakis Ah OK.  Not really sure on that one.

mfojtik commented:
@deads2k @kargakis do we want to merge this or hold? (this is a bugfix)

kargakis commented on pkg/deploy/controller/deployment/deployment_controller.go:
```diff
@@ -414,9 +428,64 @@ func (c *DeploymentController) makeDeployerContainer(strategy *deployapi.Deploym
+		oldPodBytes, err := runtime.Encode(encoder, pod)
+		if err != nil {
+			errors = append(errors, err)
+		}
```
Missing a continue?

mfojtik replied:
good catch, fixed.

kargakis commented:
> @deads2k @kargakis do we want to merge this or hold? (this is a bugfix)

I am fine with merging.

deads2k commented:
> I am fine with merging.

Likewise.  Looks like @kargakis has reviewed, so assigning him.

kargakis commented:
Yeah, I am missing the etcd endpoint, thanks.

mfojtik commented:
@kargakis @deads2k the infra policies now need re-review as I:

* renamed the deployment controller to deployer controller (the existing deployment controller SA is used by upstream deployment controller and we should not use it for what we previously called "deployment" controller.
* a new SA was added for deployer controller

kargakis commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -229,14 +232,22 @@ func init() {
 				// DeploymentControllerFactory.deploymentClient
 				{
-					Verbs:     sets.NewString("get", "update"),
+					Verbs:     sets.NewString("create", "get", "update", "delete"),
```
Squash with the above rule, the comments are obsolete anyway.

kargakis commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -250,6 +261,34 @@ func init() {
+					Resources: sets.NewString("pods"),
+				},
+				{
+					Verbs:     sets.NewString("create", "get", "list", "watch", "update", "delete"),
```
I don't think the deployer controller is creating or deleting RCs

kargakis commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -250,6 +261,34 @@ func init() {
+			Rules: []authorizationapi.PolicyRule{
+				{
+					APIGroups: []string{""},
+					Verbs:     sets.NewString("create", "get", "list", "watch", "update", "patch", "delete"),
```
This may be a minor comment but I don't see update used anywhere.

kargakis commented:
A couple of comments on the new role, rest LGTM

mfojtik commented:
@kargakis will address the auth using new way @deads2k is building up and then merge this (hopefully end if this week).

mfojtik commented:
@kargakis @deads2k refactored for the new controller initialization, updated RBAC roles, added deployment-trigger SA, renamed deployement to deployer and finally the patching for ownerRefs in deployerPods ;-)

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func eventsRule() rbac.PolicyRule {
+	return rbac.NewRule("create", "update", "patch").Groups(kapiGroup).Resources("events").RuleOrDie()
```
Did you find out why we have update and patch here? Is it event compaction?

mfojtik replied:
@deads2k @sttts ? no clue, but we have that in other controllers...

mfojtik replied:
@deads2k made a func for it so it must be useful ;-)

deads2k replied:
> Did you find out why we have update and patch here? Is it event compaction?

Yes,  it's still done client-side after all this time.

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -0,0 +1,105 @@
+	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + InfraDeploymentConfigControllerServiceAccountName},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("get", "list", "watch").Groups(kapiGroup).Resources("pods").RuleOrDie(),
```
We can remove this rule and anything related to pods in the DC controller after https://github.com/openshift/origin/pull/14046 is merged.

mfojtik replied:
sure, I will update this PR.

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -0,0 +1,105 @@
+	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + InfraDeploymentTriggerControllerServiceAccountName},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("get", "list", "watch").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
We shouldn't need to list or watch RCs here, the only thing this controller needs is get.

mfojtik replied:
kcorelistersinternal.NewReplicationControllerLister(rcInformer.GetIndexer())

will this work only with get?

kargakis replied:
List and watch are used by the informer but we have no use for the informer in the controller. We just GET rcs and compare their templates w/ the DC template.

kargakis replied:
I guess since we use the shared informer, we are bound to have list and watch for this controller?

mfojtik replied:
@kargakis that is what I think @deads2k ?

deads2k replied:
It's hard to be consistent since its not enforced, but having the full permission allows us to split the controller into a separate, unprivileged pod.  In places where we know it, I'd prefer to keep it.

There are two opinions generally held about controllers:
 1. having discrete pods for each controller with their own service account provides a very nice mental, code, and operation separation.
 1. having a single process running all controllers reduces master load, operational complexity (watch that basket!), and aggregate memory usage.

Generally speaking, a single person believes both things and can't make a choice. :)

mfojtik replied:
iow. when we run this as a standalone process in a pod, we will need these permissions to get the shared informer?

deads2k replied:
> iow. when we run this as a standalone process in a pod, we will need these permissions to get the shared informer?

Yes.  and I acknowledge that such a separation may be a long way off.

kargakis replied:
Hard choice, both options have pros and cons.

mfojtik replied:
well given that this controller was on privileged loopback client before, I would leave this as it is for now?

kargakis replied:
Yes, it's fine

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -0,0 +1,41 @@
+	// initialize build controller
+	storageVersion := c.Options.EtcdStorageConfig.OpenShiftStorageVersion
+	groupVersion := schema.GroupVersion{Group: "", Version: storageVersion}
+	codec := kapi.Codecs.LegacyCodec(groupVersion)
```
looks like everyone wants a codec.  Add a TODO to add it to our controller context.

kargakis replied:
@mfojtik TODO or issue please

mfojtik replied:
todo should be already in place.

mfojtik replied:
seems like github is not catching up ;-)

deads2k commented on pkg/cmd/server/origin/controller/apps.go:
```diff
@@ -0,0 +1,90 @@
+func (c *DeployerControllerConfig) RunController(ctx ControllerContext) (bool, error) {
+	internalDeployerKubeClient, err := ctx.ClientBuilder.KubeInternalClient(bootstrappolicy.InfraDeployerControllerServiceAccountName)
+	if err != nil {
+		return false, err
```
`return true, err` . You were enabled and you failed.  Returning false means you had some secondary bit of data that said "I don't need to start".  Discovery revealing that your resources were missing perhaps?  A check for  a later time I think.

kargakis commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -699,20 +699,66 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
+	allowedOpenshiftControllers := sets.NewString(
+		"deployer",
+		"deploymentconfig",
+		"deploymenttrigger",
```
Where are these names coming from?

mfojtik replied:
names of the controllers... eventually we will probably create wrapper for apps.RunController (run all controllers in apps group because users don't want to disable just one deployment controller)

mfojtik replied:
@kargakis any objections from using these names?

kargakis replied:
I am fine with the names

mfojtik replied:
@deads2k will it make more sense to prefix these names with group? iow. `apps.openshift.io/controllers/deployer` or something

deads2k replied:
> @deads2k will it make more sense to prefix these names with group? iow. apps.openshift.io/controllers/deployer or something...

Probably, but leave it for another pull.  I don't want this one delayed on bikeshedding we can sort out next week.

mfojtik replied:
+1 there are many controllers to go ;-)

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -994,23 +1009,27 @@ func (c *MasterConfig) DeploymentConfigInstantiateClients() (*osclient.Client, k
-func (c *MasterConfig) DeploymentControllerClients() (*osclient.Client, kclientsetinternal.Interface, kclientsetexternal.Interface) {
-	_, osClient, internalKubeClientset, externalKubeClientset, err := c.GetServiceAccountClients(bootstrappolicy.InfraDeploymentConfigControllerServiceAccountName)
+// DeploymentConfigClients returns deploymentConfig and deployment client objects
+func (c *MasterConfig) DeploymentConfigClients() (*osclient.Client, kclientsetinternal.Interface) {
```
who uses these now?

mfojtik replied:
good catch, should be nuked.

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -994,23 +1009,27 @@ func (c *MasterConfig) DeploymentConfigInstantiateClients() (*osclient.Client, k
+}
+
+// DeployerControllerClients returns the deployer controller client objects
+func (c *MasterConfig) DeployerControllerClients() (*osclient.Client, kclientsetinternal.Interface, kclientsetexternal.Interface) {
```
who uses these now?

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -994,23 +1009,27 @@ func (c *MasterConfig) DeploymentConfigInstantiateClients() (*osclient.Client, k
-}
-
 // DeploymentConfigControllerClients returns the deploymentConfig controller client objects
 func (c *MasterConfig) DeploymentConfigControllerClients() (*osclient.Client, kclientsetinternal.Interface, kclientsetexternal.Interface) {
```
who uses these now?

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -699,20 +699,66 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
+		started, err := initFn(openshiftControllerContext)
 		if err != nil {
-			glog.Fatalf("Could not start build controller: %v", err)
+			glog.Errorf("Error starting %q", controllerName)
```
Huh, I thought I made this process death.  Should fatal here I think to allow us to crashloop.

mfojtik commented:
(need to wait for https://github.com/openshift/origin/pull/14046 to land before merging this)

kargakis commented:
> (need to wait for #14046 to land before merging this)

No need to block on it, dropping the pod-related permissions can be a follow-up

mfojtik commented:
@kargakis in that case this should be ready to go now.

mfojtik commented:
[test]

mfojtik commented:
@deads2k fully rebased (including your stuff) and as bonus i plumbed @smarterclayton new controller for new initialization and rbac rules.

we should ship this asap to avoid future conflics and also I want to stop this PR from growing ;-)

kargakis commented on pkg/deploy/controller/deployment/deployment_controller_test.go:
```diff
@@ -436,8 +439,9 @@ func TestHandle_noop(t *testing.T) {
 
-		if len(client.Actions()) > 0 {
-			t.Errorf("%s: unexpected actions: %v", test.name, client.Actions())
+		// Expect only patching for ownerRefs
```
Check that we are indeed patching

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -92,6 +92,21 @@ func init() {
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("list", "watch").Groups(imageGroup, legacyImageGroup).Resources("imagestreams").RuleOrDie(),
+			rbac.NewRule("get", "update").Groups(extensionsGroup).Resources("daemonsets").RuleOrDie(),
+			rbac.NewRule("get", "update").Groups(extensionsGroup, appsGroup).Resources("deployments").RuleOrDie(),
```
missing statefulsets?

mfojtik replied:
good catch

deads2k commented:
lgtm



> we should ship this asap to avoid future conflics and also I want to stop this PR from growing ;-)

I think you're screwed no matter what.  The queue is still in double digits and every rebase puts you back to the back.

mfojtik commented:
@deads2k unless I do this Clayton way and merge it manually ;-)

mfojtik commented:
@kargakis all fixed, can we start shipping this?

kargakis commented:
lgtm but integration tests for builds seem broken

deads2k commented:
[merge]

mfojtik commented:
@deads2k rolled back the changes to ICT controller, will open a separate PR for that once this merge.

mfojtik commented:
@bparees @coreydaley pls check, I made some tweaks to accomodate the cleanup policy (i think the build controller now needs ability to delete builds and create build configs?)

bparees commented:
> @bparees @coreydaley pls check, I made some tweaks to accomodate the cleanup policy (i think the build controller now needs ability to delete builds and create build configs?)

@mfojtik looks like this merged before we had a chance to look at it...    why would the build controller need the ability to create buildconfigs?

kargakis commented:
> @mfojtik looks like this merged before we had a chance to look at it... why would the build controller need the ability to create buildconfigs?

If you delete something with an owner ref, you have to be able to create owners. I think.

kargakis commented:
> If you delete something with an owner ref, you have to be able to create owners. I think.

No, I confused this with another rule: adding owner refs to things means you can also delete them. Not sure about this one.

mfojtik commented:
sorry it was not create but delete build and get bc (https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/controller/build.go#L59)

bparees commented:
ok yeah those make sense.

kargakis commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -428,7 +429,10 @@ func (c *DeploymentConfigController) cleanupOldDeployments(existingDeployments [
 
-		err := c.rn.ReplicationControllers(deployment.Namespace).Delete(deployment.Name, nil)
+		policy := metav1.DeletePropagationBackground
+		err := c.rn.ReplicationControllers(deployment.Namespace).Delete(deployment.Name, &metav1.DeleteOptions{
```
@tnozicka fyi this is related to DeploymentConfig GC
